MapboxNavigation/NavigationMapView.swift

```diff
@@ -1136,6 +1144,57 @@ open class NavigationMapView: MGLMapView, UIGestureRecognizerDelegate {
         return MGLShapeCollectionFeature(shapes: altRoutes + [mainRoute])
     }
     
+    /**
+     Returns an array of `RoadClasses` for specific leg.
+     
+     - parameter leg: Leg of the route.
+     - returns: A list of `RoadClasses` for specific `RouteLeg`. `RoadClasses` will be set to `nil` if it's not present in `Intersection`.
+     */
+    func roadClasses(_ leg: RouteLeg) -> [RoadClasses?] {
+        // Iterate over `RouteStep`s and pick only valid segment indices for specific `Intersection`.
+        // Array of segment indexes can look like this: [0, 3, 24, 28, 48, 50, 51, 53].
+        var segmentIndicesByIntersection = [Int]()
+        leg.steps.forEach {
+            if let segmentIndices = $0.segmentIndicesByIntersection?.compactMap({ $0 }) {
+                segmentIndicesByIntersection.append(contentsOf: segmentIndices)
+            }
+        }
+        
+        // Iterate over each `Intersection` and save `RoadClasses`.
+        // Array of `RoadClasses` can look like this:
+        // [Optional(toll,motorway), ... , Optional(), Optional(toll,motorway), nil]
+        var roadClassesInLeg = [RoadClasses?]()
+        leg.steps.forEach {
+            $0.intersections?.forEach {
+                var roadClass = $0.outletRoadClasses
+                if let roadClasses = roadClass, roadClasses.isEmpty {
+                    roadClass = nil
+                }
+                
+                roadClassesInLeg.append(roadClass)
+            }
+        }
```

Comment:
We could write this more functionally by flat-mapping legs to steps, flat-mapping steps to intersections, mapping intersections to road classes, and filtering out empty road classes.

Reply:
Compact map in second step is not needed as we need to retain `nil`s. In any case thanks for suggestions, I've modified implementation to use higher order functions.